src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs

```diff
@@ -321,6 +327,348 @@ public async Task TestRemoteWorkspaceCircularReferences()
             Assert.NotNull(solution);
         }
 
+        private static ImmutableArray<ImmutableArray<T>> Permute<T>(T[] values)
+        {
+            using var _ = ArrayBuilder<ImmutableArray<T>>.GetInstance(out var result);
+            DoPermute(0, values.Length - 1);
+            return result.ToImmutable();
+
+            void DoPermute(int start, int end)
+            {
+                if (start == end)
+                {
+                    // We have one of our possible n! solutions,
+                    // add it to the list.
+                    result.Add(values.ToImmutableArray());
+                }
+                else
+                {
+                    for (var i = start; i <= end; i++)
+                    {
+                        (values[start], values[i]) = (values[i], values[start]);
+                        DoPermute(start + 1, end);
+                        (values[start], values[i]) = (values[i], values[start]);
+                    }
+                }
+            }
+        }
+
+        private static async Task TestInProcAndRemoteWorkspace(
+            bool syncWithRemoteServer, params ImmutableArray<(string hintName, SourceText text)>[] values)
+        {
+            // Try every permutation of these values.
+            foreach (var permutation in Permute(values))
+            {
+                var gotException = false;
+                try
+                {
+                    await TestInProcAndRemoteWorkspaceWorker(syncWithRemoteServer, permutation);
+                }
+                catch (XunitException)
+                {
+                    gotException = true;
+                }
+
+                // If we're syncing to the remove server, we should get no exceptions, since the data should be matched
+                // between both.  If we're not syncing, we should see a failure since the two processes will disagree on
+                // the contents.
+                Assert.Equal(syncWithRemoteServer, !gotException);
```

Comment:
You might want to change this to say something like:

    if (!syncWithRemoteServer)
        Assert.Throws(() => .....)
    else
        // call method directly

Might give you a slightly nicer failure message than "expected true, got false" or something. (and is also just clearer on intent)